nibabel/funcs.py

```diff
@@ -224,7 +224,18 @@ def as_closest_canonical(img, enforce_diag=False):
     # we need to transform the data
     arr = img.get_data()
     t_arr = apply_orientation(arr, ornt)
-    return img.__class__(t_arr, out_aff, img.header)
+    # Also apply the transform to the dim_info fields
+    new_hdr = img.header.copy()
```

Comment:
Once I get agreement that the rest of the comments make sense and you're happy with the general idea, I can try and implement this.